### ISSUE TITLE
no-unused-vars false positive for use on right side of assignment

### ISSUE THREAD

**Issue by btmills** [eslint/eslint#7282]
**Tell us about your environment**

* **ESLint Version:** http://eslint.org/demo
* **Node Version:** n/a
* **npm Version:** n/a

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

None, inline

**What did you do? Please include the actual source code causing the issue.**

```js
/* eslint no-unused-vars: "error" */
// 'foo' is defined but never used. (no-unused-vars)
var foo;
foo = 42;
foo = foo();
foo = foo.bar();
foo = bar(foo);

// No error
var bar;
bar = 42;
bar();
```

**What did you expect to happen?**

No error. `foo` is used multiple times on the right side of the assignment expressions.

**What actually happened? Please include the actual, raw output from ESLint.**

3:5 - 'foo' is defined but never used. (no-unused-vars)

It looks like references are being excluded on both sides of assignment expressions instead of just the left side.

Originally seen on http://stackoverflow.com/questions/39787723/eslint-no-unused-vars-stream-conditional-pipe-use-case

**Comment by not-an-aardvark** [comment]
Working on this.

**Comment by not-an-aardvark** [comment]
Actually, looking at this again, is it actually a bug? The assignment in line 7 is unnecessary, since `foo` is never used afterwards. (The line could be replaced with just `bar(foo);` and the rule would pass.)

I suppose it is a bit misleading to report an error on line 3 instead of line 7, though.

**Comment by btmills** [comment]
The left side of the final assignment expression is unnecessary, but the variable is "used", at least according to my interpretation of the rule. You can look at the linked StackOverflow question for a less trivial example. In an ideal world (and I believe this is what you were getting at when you suggested warning on line 7 instead of the declaration), we'd be able to determine statically that the result of the final assignment expression is unused and suggest using the right side by itself. Even supposing that were possible (it may be, and that be really cool), I don't think that warning should come from no-unused-vars. Put another way, my interpretation of no-unused-vars is that when it declares a variable as unused, I ought to be able to remove that variable from my program without changing the program's behavior. Does that fit?

Also, welcome to the team! :tada:

**Comment by not-an-aardvark** [comment]
I think the current version of the rule *does* try to determine statically that the last assignment is unused, and that's why it reports an error in this case. (In the StackOverflow question, the line `stream = stream.pipe(gulp.dest(manifestRoot));` can be replaced with `stream.pipe(gulp.dest(manifestRoot));` without changing anything)

Perhaps the name `no-unused-vars` is a bit misleading (could be `no-unused-assignments` instead, but I think the ability to lint for unused assignments is very useful. In lieu of creating another rule to check for unused assignments, I think it would be better to keep this error and improve the error message (i.e. ensure that it's on the right line, and that the message more accurately describes the issue).

**Comment by mysticatea** [comment]
This was enhanced in #6348.

**Comment by mysticatea** [comment]
Also, there is a document issue: #6703

**Comment by not-an-aardvark** [comment]
So I propose we make the following changes:

* We should report every case where an unused assignment happens. (At the moment, we only report it in one place -- where the variable was declared.)
* We should improve the error message in cases like this (e.g. `Unused assignment to variable 'foo'`).

**Comment by nzakas** [comment]
> We should report every case where an unused assignment happens. (At the moment, we only report it in one place -- where the variable was declared.)

:-1: That will be very noisy and I don't think it adds much to the user experience. This rule is not about assignments, it's about identifiers that were defined and then never used. We do not consider an assignment only to be used, so this should warn:

@mysticatea you removed "accepted" from this but didn't indicate why. Can you explain?

**Comment by not-an-aardvark** [comment]
@nzakas I see your point about which line to report. However, I still think we should improve the error in these cases (even if just by improving the error message when a variable is assigned but never read, like what happens in the following case):

```js
var foo;

// other logic

foo = 5;
```

This has led to a lot of confusion for users (see https://github.com/eslint/eslint/issues/3417, https://github.com/eslint/eslint/issues/3129, https://github.com/eslint/eslint/issues/3719, https://github.com/eslint/eslint/issues/3917, https://github.com/eslint/eslint/issues/4830, https://github.com/eslint/eslint/issues/5177, https://github.com/eslint/eslint/issues/5727, https://github.com/eslint/eslint/issues/5942, https://github.com/eslint/eslint/issues/6243, https://github.com/eslint/eslint/issues/7027, https://github.com/eslint/eslint/issues/7107, https://github.com/eslint/eslint/issues/7286). The confusion that is that users think `foo` is foo is used because it's assigned to, even though the assignment is unnecessary. The rule is certainly correct in reporting these cases, but maybe we should make it more obvious that the rule took the assignment into account.

**Comment by mysticatea** [comment]
@nzakas I thought this is the intentional behavior of #6348 and duplicate of #6703. What do I overlook?

**Comment by platinumazure** [comment]
I'm not sure that the point of the rule is to detect that the most recent assignment of a variable is useless, if the variable was used (on an RHS, and not as part of a self-assign) before that assignment.

Seems we should be able to use code path analysis to determine if a variable has been used before being assigned to.

Am I missing something?

**Comment by nzakas** [comment]
@not-an-aardvark I'm open to changing the message that's reported in order to avoid confusion in those cases. Maybe something like "foo is assigned a value but is never used" instead?

@mysticatea no need to apologize. It's just that you didn't leave a comment explaining why you removed "accepted," so I wasn't sure if you disagreed that this was a bug, or something else.
